test/SignatureValidator/EIP712.t.sol

```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.7.6;
+pragma abicoder v2;
+
+import { TestSignatureValidator } from "./Setup.t.sol";
+
+contract TestEIP712 is TestSignatureValidator {
+    uint8 public constant sigType = uint8(SignatureType.EIP712);
+
+    function testEIP712WithDifferentSigner() public {
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(otherPrivateKey, digest);
+        bytes memory signature = abi.encodePacked(r, s, v, sigType);
+        assertFalse(isValidSignature(vm.addr(userPrivateKey), digest, bytes(""), signature));
+    }
+
+    function testEIP712WithWrongHash() public {
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(userPrivateKey, digest);
+        bytes memory signature = abi.encodePacked(r, s, v, sigType);
+        bytes32 otherDigest = keccak256("other data other data");
+        assertFalse(isValidSignature(vm.addr(userPrivateKey), otherDigest, bytes(""), signature));
+    }
+
+    function testEIP712WithWrongSignatureLength() public {
+        uint256 r = 1;
+        bytes memory signature = abi.encodePacked(r, sigType);
+        // should have 33 bytes signature
+        assertEq(signature.length, 33);
+        vm.expectRevert("SignatureValidator#isValidSignature: length 65 or 97 required");
+        isValidSignature(vm.addr(userPrivateKey), digest, bytes(""), signature);
+    }
+
+    /// @dev old contracts still assert sigLength == 97 so has to support this format
+    /// @dev the extra bytes32 is not used at all
+    function testEIP712With97BytesSig() public {
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(userPrivateKey, digest);
+        bytes memory signature = abi.encodePacked(r, s, v, bytes32(0), sigType);
+        // standard ECDSA signature : 65 bytes
+        // extra bytes32 : 32 bytes
+        // signatureType : 1 byte
+        // total : 98 bytes
+        assertEq(signature.length, 98);
```

Comment:
Q: So the signature length should be 97 (mentioned in test name) or 98? 😂

Reply:
Hmm it depends on the definition of signature. The total length of `signature` passing into function `isValidSignature()` is 98 bytes (97 + 1 for signature type). I will change the name to 98 to avoid misunderstanding.

Reply:
Fixed a978a7b2ece6e54f6cadc2fb20c3623625dd806f